src/go/networking/pkg/portproxy/server.go

```diff
@@ -73,47 +83,139 @@ func (p *PortProxy) handleEvent(conn net.Conn) {
 		logrus.Errorf("port server decoding received payload error: %s", err)
 		return
 	}
-	p.execListener(pm)
+	p.exec(pm)
 }
 
-func (p *PortProxy) execListener(pm types.PortMapping) {
-	for _, portBindings := range pm.Ports {
-		for _, portBinding := range portBindings {
-			logrus.Debugf("received the following port: [%s] from portMapping: %+v", portBinding.HostPort, pm)
-			port, err := nat.ParsePort(portBinding.HostPort)
-			if err != nil {
-				logrus.Errorf("parsing port error: %s", err)
-				continue
-			}
-			if pm.Remove {
-				p.mutex.Lock()
-				if listener, exist := p.activeListeners[port]; exist {
-					logrus.Debugf("closing listener for port: %d", port)
-					if err := listener.Close(); err != nil {
-						logrus.Errorf("error closing listener for port [%s]: %s", portBinding.HostPort, err)
-					}
+func (p *PortProxy) exec(pm types.PortMapping) {
+	for portProto, portBindings := range pm.Ports {
+		logrus.Debugf("received the following port: [%s] and protocol: [%s] from portMapping: %+v", portProto.Port(), portProto.Proto(), pm)
+		if gvisorTypes.TransportProtocol(portProto.Proto()) == gvisorTypes.TCP {
+			p.handleTCP(portBindings, pm.Remove)
+		}
+		if gvisorTypes.TransportProtocol(portProto.Proto()) == gvisorTypes.UDP {
+			p.handleUDP(portBindings, pm.Remove)
+		}
+	}
+}
+
+func (p *PortProxy) handleUDP(portBindings []nat.PortBinding, remove bool) {
+	for _, portBinding := range portBindings {
+		port, err := nat.ParsePort(portBinding.HostPort)
+		if err != nil {
+			logrus.Errorf("parsing port error: %s", err)
+			continue
+		}
+		if remove {
+			p.udpConnMutex.Lock()
+			if udpConn, exist := p.activeUDPConns[port]; exist {
+				logrus.Debugf("closing UDPConn for port: %d", port)
```

Comment:
`Debugf` with a mutex held seems kind of heavy?

Reply:
I agree, however, since it's only going to output when debugging is on, do you still think that would an issue?

Reply:
There's also a potential `Errorf` later. But yeah, it's probably fine.